Add Write Models Docs (#203)
Oneflow-Inc/libai: docs/source/tutorials/Write_Models.md

# Write Models

> In this section, we will introduce how to implement new model entirely from scratch and make it compatible with LiBai.

Reviewer: ```suggestion
In this section, we will introduce how to implement a new model entirely from scratch and make it compatible with LiBai.
```

Author: 已改~

> In this section, we are showing users how to implement new model entirely from scratch and make it compatible with LiBai.

Reviewer 2: ```suggestion
In this section, we will introduce how to implement new model entirely from scratch and make it compatible with LiBai.
```

Author: 已改~

> Here we provides the tutorials for users who want to implement new model entirely from scratch and make it compatible in LiBai.

Reviewer 3: ```suggestion
In this section, we are showing users of how to implement new model entirely from scratch and make it compatible with LiBai.
```

Reviewer 3: 这样如何

## Construct Models in LiBai

> LiBai uses ``LazyConfig`` for more flexible config system, which means you can simply import your own model in your config and train it under LiBai.

Reviewer 2: 这里的lazyconfig可以给个 超链接到lazyconfig的doc说明

Author: 好，OK

> For image classification task, the input data is usually a batch of images and labels during training, and only images when testing. The following code shows how to build a toy model for image classification task, import this code in your code:

Reviewer 2: 这里可能不太准确, 因为在libai里面  在`testing`的时候 也是有`labels`的

Author: OK，我来改一下

Author: 我改了~再看看呢~

> For example, in image classification task, the input data is usually a batch of images with its targets during training and only input a batch of images when testing, to build a toy model for image classification task, import this code in your own code:

Reviewer 3: ```suggestion
For image classification task, the input data is usually a batch of images and labels during training, and only images when testing. The following code shows how to build a toy model for image classification task:
```

> For image classification task, the input data is usually a batch of images and labels. The following code shows how to build a toy model for image classification task, import this code in your code:

Reviewer: ```suggestion
For image classification task, the input data is usually a batch of images and labels. The following code shows how to build a toy model for this task, import it in your code:
```

Author: 已改~

```python
# toy_model.py
import oneflow as flow
import oneflow.nn as nn

class ToyModel(nn.Module):
    def __init__(self, 
        num_classes=1000, 

>         loss_func=None

Reviewer: 这里 loss_func 作为入参，并没有用到

Author: OK, 我这边删了~

    ):
        super().__init__()
        self.features = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1)
        self.avgpool = nn.AdaptiveAvgPool2d(1)
        self.classifier = nn.Linear(64, num_classes)
        self.loss_func = nn.CrossEntropyLoss()
    
    def forward(self, images, labels=None):
        x = self.features(images)
        x = self.avgpool(x)
        x = flow.flatten(x, 1)
        x = self.classifier(x)

        if labels is not None and self.training:
            losses = self.loss_func(x, labels)
            return {"losses": losses}
        else:
            return {"prediction_scores": x}
```

**Note:**

> - For classification models, the ``forward`` function must have ``images`` and ``labels`` as arguments, which corresponds to the output of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.

Reviewer: ```suggestion
- For classification models, the ``forward`` function must have ``images`` and ``labels`` as arguments, which corresponds to the output in ``getitem`` of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.
```

Author: 已改~，我改为`__getitem__`

> - For classification models, the ``forward`` function of each model must have ``images`` and ``labels`` as arguments, which corresponds to the output of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.

Reviewer 3: ```suggestion
- For classification models, the ``forward`` function must have ``images`` and ``labels`` as arguments, which corresponds to the output of LiBai's built-in datasets, please refer to [imagenet.py](https://github.com/Oneflow-Inc/libai/blob/main/libai/data/datasets/imagenet.py) for more details about the dataset.
```

Author: 已修改~

> - Each model will return ``losses`` during training and ``prediction_scores`` during inference, both of which should be the type of ``dict``, which means you should implement the ``loss function`` in your model, like ``self.loss_func=nn.CrossEntropyLoss()`` as the ToyModel showing above.

Reviewer: ```suggestion
- This toy model will return ``losses`` during training and ``prediction_scores`` during inference, and both of them should be the type of ``dict``, which means you should implement the ``loss function`` in your model, like ``self.loss_func=nn.CrossEntropyLoss()`` as the ToyModel showing above.
```

这里强调一下是这个 example 里面的 toy model，不然用户会以为所有的 model 都需要返回这些内容，有的模型在 inference 的时候可以返回更多的内容

Author: OK，已修改~

> - Each model returns ``losses`` during training and returns ``prediction_scores`` during inference, and both of them should be the type of ``dict``, which means you should implement the ``loss function`` in your model, like ``self.loss_func=nn.CrossEntropyLoss()`` as the ToyModel.

Author: 已修改~

## Import the model in config

> With ``LazyConfig System``, you can simply import the model in your config file to use ``ToyModel`` for training as follows:

Reviewer 3: ```suggestion
With ``LazyConfig System``, you can simply import the model in your config file. The following code shows how to use ``ToyModel`` for training:
```

Author: 已修改~

> With ``LazyConfig System``, you can simply import the model in your config file. The following code shows how to import ``ToyModel`` in your config file:

Reviewer: ```suggestion
With ``LazyConfig System``, you can simply import the model in your config file. The following code shows how to use ``ToyModel`` in your config file:
```

Author: 已改~
